managed/services/onboarding/tips.go

```diff
@@ -0,0 +1,285 @@
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package onboarding provides functionality for user onboarding features.
+package onboarding
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/api/onboardingpb"
+	"github.com/percona/pmm/managed/models"
+)
+
+const (
+	// system tip ids
+	InstallPMMServerTipID    = 1
+	InstallPMMClientTipID    = 2
+	ConnectServiceToPMMTipID = 3
+
+	// user tip ids
+	SeeYourNewAdvisorsTipID = 1000
+)
+
+type TipsService struct {
+	db               *reform.DB
+	inventoryService inventoryService
+	c                grafanaClient
+
+	systemTipIDs map[int64]struct{}
+	userTipIDs   map[int64]struct{}
+
+	onboardingpb.UnimplementedTipServiceServer
+}
+
+var _ onboardingpb.TipServiceServer = (*TipsService)(nil)
+
+func NewTipService(db *reform.DB, inventoryService inventoryService, client grafanaClient) *TipsService {
+	return &TipsService{
+		db:               db,
+		inventoryService: inventoryService,
+		c:                client,
+		systemTipIDs: map[int64]struct{}{
+			InstallPMMServerTipID:    {},
+			InstallPMMClientTipID:    {},
+			ConnectServiceToPMMTipID: {},
+		},
+		userTipIDs: map[int64]struct{}{
+			SeeYourNewAdvisorsTipID: {},
+		},
+	}
+}
+
+func (t *TipsService) GetOnboardingStatus(ctx context.Context, _ *onboardingpb.GetOnboardingStatusRequest) (*onboardingpb.GetOnboardingStatusResponse, error) {
+	userID, err := t.c.GetUserID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	systemTips, err := t.retrieveSystemTips()
+	if err != nil {
+		return nil, err
+	}
+
+	userTips, err := t.retrieveUserTips(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &onboardingpb.GetOnboardingStatusResponse{
+		SystemTips: systemTips,
+		UserTips:   userTips,
+	}, nil
+}
+
+func (t *TipsService) retrieveSystemTips() ([]*onboardingpb.TipModel, error) {
+	structs, err := t.db.Querier.SelectAllFrom(models.OnboardingSystemTipTable, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve system tip by id")
+	}
+	tips := make([]*models.OnboardingSystemTip, len(structs))
+	for i, s := range structs {
+		tips[i] = s.(*models.OnboardingSystemTip) //nolint:forcetypeassert
+	}
+
+	for _, tip := range tips {
+		if !tip.IsCompleted {
+			isCompleted, err := t.isSystemTipCompleted(tip.ID)
+			if err != nil {
+				return nil, err
+			}
+			tip.IsCompleted = isCompleted
+
+			err = t.db.Save(tip)
```

Comment:
Do we need to save the tip, if its current `isCompleted` value is the same as the new one?

Another question: should a method called `retrieveSystemTips` mutate records at all?

Reply:
It won't update the tip status in DB if isCompleted is true, it would check it if isCompleted is false.

Refactored code.